akka-actor-typed/src/main/scala/akka/actor/typed/StatusResponse.scala

```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.actor.typed
+
+import akka.annotation.DoNotInherit
+
+/**
+ * Top level message type for signalling responses that may fail or succeed. Has special support in akka-actor-typed ask.
+ *
+ * Not for user extension.
+ *
+ * @tparam T the wrapped successful response type
+ */
+@DoNotInherit
+sealed trait StatusResponse[+T]
+
+object StatusResponse {
+  final case class Ok[T](value: T) extends StatusResponse[T]
+  final case class Fail(errorMessage: String) extends StatusResponse[Nothing]
```

Comment:
I think we can make it `Fail(cause: Throwable)` and it can propagate the exception if it can be serialized.

The `errorMessage: String` case is very common and we can have a built in exception type for that:
```
final case class ErrorMessage(message: String) extends RuntimeException(message) with NoStackTrace
```

The the Future would be failed with the `ErrorMessage` instead of the too generic `RuntimeException`.
We would also have convenience apply
```
object Fail {
  def apply(errorMessage: String): Fail = apply(ErrorMessage(errorMessage))
}
```

Reply:
We very often use the work 'reply' when talking about 'ask' interactions. I think `Reply` better in that case. 

But I would even argue that `Result` is a better name. True that we plain to use it for `ask`, but in its essence, we are introducing a more opinionated version of `Either` that will also play well with Java. Hence, `Result`, `Ok` and `Fail`. 

We could then have:

```scala
sealed trait Result[+T]
final case class Ok[T](value: T) extends Result[T]
final case class Fail(errorMessage: String) extends Result[Nothing]
```

I'm not sure about making it a `Throwable` though. I think there is value in avoiding it. If it's a throwable, people can easily throw it. The fact that we could eventually convert it to a throwable internally (when crossing the boundary) doesn't imply it must be a throwable since the beginning.

Reply:
Btw, any reason for adding `Ok` and `Fail` inside the companion of the parent trait?

Reply:
I'd also prefer `Reply`.

Another consideration is if we should have the two types `Ok`, `Fail`? Since it's not supposed to be used in pattern match an alternative would be to use something like:
```scala
object Reply {
  def apply[A](value: A): Reply[A] =
    new Reply(Success(value))

  def of[A](value: A): Reply[A] =
    apply(value)

  def ofError[A](error: Throwable): Reply[A] =
    new Reply(Failure(error))

  def ofError[A](errorMessage: String): Reply[A] =
    ofError(ErrorMessage(errorMessage))

  final case class ErrorMessage(message: String) extends RuntimeException(message) with NoStackTrace
}

final case class Reply[A] private (tryValue: Try[A]) {
  def getValue: A = tryValue.get

  def getError: Throwable = tryValue match {
    case Success(_)   => throw new IllegalArgumentException("Expected reply to be error, but was success.")
    case Failure(exc) => exc
  }

  def isSuccess: Boolean = tryValue.isSuccess

  def isError: Boolean = tryValue.isFailure
}
```

Might look somewhat better when sending the Reply for the normal case:
```
replyTo ! Reply(17)
```

We could ofc have such Reply.apply also if we keep `Ok`/`Fail`.

Reply:
Why aren't we just using `Try`? Because of Java?

Reply:
While it is concise I'm not sold on `Reply` because that convene anything about how it is different from any other reply. I think that needs to go in the type, that could still contain `Reply` though. So some combination with `Failable` or `Status` would be good IMO.

I like hiding the concrete adt types behind factories like in Patriks suggestion, let's do that.

I think providing a nice API for Java is a good motivation for not using `Try`, not encouraging  sending standard library types I think is another (but less strong) argument.

Reply:
hm, difficult naming. I'm not enthusiastic about Status but can't come up with something better.

`askStatus` - `ReplyStatus.ok(...)`, `ReplyStatus.err(...)`

`askWithStatus` - `ReplyWithStatus.ok(...)`, `ReplyWithStatus.err(...)`

I agree that we should have a more specific type than sending `Try`.

Reply:
I think `askWithStatus` / `ReplyWithStatus` is the least bad so far, I'll move forward with that.